pkg/collector/corechecks/windows_event_log/check.go

```diff
@@ -0,0 +1,284 @@
+
+	fetchEventsLoopWaiter sync.WaitGroup
+	fetchEventsLoopStop   chan struct{}
+
+	includedMessages []*regexp.Regexp
+	excludedMessages []*regexp.Regexp
+
+	// event metrics
+	eventPriority agentEvent.EventPriority
+
+	// event log
+	session             evtsession.Session
+	sub                 evtsubscribe.PullSubscription
+	evtapi              evtapi.API
+	systemRenderContext evtapi.EventRenderContextHandle
+	userRenderContext   evtapi.EventRenderContextHandle
+	bookmarkSaver       *bookmarkSaver
+}
+
+// Run executes the check
+func (c *Check) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+	// Necessary for check stats to be calculated (number of events collected, etc)
+	defer sender.Commit()
+
+	if !c.sub.Running() {
+		err := c.startSubscription()
+		if err != nil {
+			return err
+		}
+	}
+
+	// save/persist bookmark on each check run
+	// check.Cancel() may not be called (like in agent check command), which means
+	// if the events read are less than bookmark_frequency then a bookmark won't
+	// be saved before the process exits. saving here, too, gives us a good time periodic
+	// save/persist in addition to the count periodic bookmark_frequency option.
+	err = c.bookmarkSaver.saveLastBookmark()
+	if err != nil {
+		c.Warnf(err.Error())
+	}
+
+	return nil
+}
+
+func (c *Check) fetchEventsLoop(sender sender.Sender) {
+	defer c.fetchEventsLoopWaiter.Done()
+
+	// Always stop the subscription when the loop ends.
+	// The check will start the subscription and this loop again next time it runs.
+	defer c.sub.Stop()
+
+	// Save the bookmark at the end of the loop, regardless of the bookmarkFrequency
+	defer func() {
+		err := c.bookmarkSaver.saveLastBookmark()
+		if err != nil {
+			c.Warnf(err.Error())
+		}
+	}()
+
+	var addBookmarkToSubOnce sync.Once
+
+	// Fetch new events
+	for {
+		select {
+		case <-c.fetchEventsLoopStop:
+			return
+		case events, ok := <-c.sub.GetEvents():
+			if !ok {
+				// The channel is closed, this indicates an error or that sub.Stop() was called
+				// Use sub.Error() to get the error, if any.
+				err := c.sub.Error()
+				if err != nil {
+					log.Errorf("event subscription stopped with error: %v", err)
+				}
+				return
+			}
+			for _, event := range events {
+				// Submit Datadog Event
+				_ = c.submitEvent(sender, event)
+
+				// bookmarkSaver manages whether or not to save/persist the bookmark
+				err := c.bookmarkSaver.updateBookmark(event)
+				if err != nil {
+					c.Warnf(err.Error())
+				} else {
+					// If we don't have a bookmark when we create the subscription we have
+					// to add it later once we've updated it at least once.
+					addBookmarkToSubOnce.Do(func() {
+						c.sub.SetBookmark(c.bookmarkSaver.bookmark)
+					})
+				}
+
+				// Must close event handle when we are done with it
+				evtapi.EvtCloseRecord(c.evtapi, event.EventRecordHandle)
+			}
+		}
+	}
+}
+
+func (c *Check) submitEvent(sender sender.Sender, event *evtapi.EventRecord) error {
```

Comment:
Would it make sense to remove the error return type? This function does not return any errors.